README.md

[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![Signed
by](https://img.shields.io/badge/Keybase-Verified-brightgreen.svg)](https://keybase.io/hrbrmstr)
![Signed commit
%](https://img.shields.io/badge/Signed_Commits-100%25-lightgrey.svg)
[![Linux build
Status](https://travis-ci.org/hrbrmstr/attackerkb.svg?branch=master)](https://travis-ci.org/hrbrmstr/attackerkb)  
![Minimal R
Version](https://img.shields.io/badge/R%3E%3D-3.2.0-blue.svg)
![License](https://img.shields.io/badge/License-MIT-blue.svg)

# attackerkb

Tools to Query the Rapid7 AttackerKB API

## Description

Rapid7 manages a service — <https://attackerkb.com/> - where experts can
evaluate various aspects of emergent or existing vulnerabilities and the
community can query and retrieve results. Tools are provided to query
the AttackerKB API.

## What’s Inside The Tin

The following functions are implemented:

  - `attackerkb_api_key`: Get or set ATTACKERKB\_API\_KEY value
  - `kb_assessments`: Helpers to query AttackerKB assessments
  - `kb_contributors`: Helpers to query AttackerKB contributors
  - `kb_topics`: Helpers to query AttackerKB topics

## Installation

``` r
install.packages("attackerkb", repos = c("https://cinc.rud.is", "https://cloud.r-project.org/"))
# or
remotes::install_git("https://git.rud.is/hrbrmstr/attackerkb.git")
# or
remotes::install_git("https://git.sr.ht/~hrbrmstr/attackerkb")
# or
remotes::install_gitlab("hrbrmstr/attackerkb")
# or
remotes::install_bitbucket("hrbrmstr/attackerkb")
# or
remotes::install_github("hrbrmstr/attackerkb")
```

NOTE: To use the ‘remotes’ install options you will need to have the
[{remotes} package](https://github.com/r-lib/remotes) installed.

## Usage

``` r
library(attackerkb)

# current version
packageVersion("attackerkb")
## [1] '0.1.0'
```

## attackerkb Metrics

| Lang | \# Files | (%) | LoC |  (%) | Blank lines |  (%) | \# Lines |  (%) |
| :--- | -------: | --: | --: | ---: | ----------: | ---: | -------: | ---: |
| R    |        9 | 0.9 | 302 | 0.97 |          96 | 0.86 |      288 | 0.91 |
| Rmd  |        1 | 0.1 |   8 | 0.03 |          15 | 0.14 |       28 | 0.09 |

## Code of Conduct

Please note that this project is released with a Contributor Code of
Conduct. By participating in this project you agree to abide by its
terms.


NEWS.md
0.1.2
* Fixed bug in the post-request processor for `/topic` API request.
  (Reported by @kwlin-r7)

0.1.1
* Fixed an issue where errors wld be thrown if zero API results came back
* Added progress spinner for API calls vs the hacky dots

0.1.0 
* Initial release


tests/tinytest.R

if ( requireNamespace("tinytest", quietly=TRUE) ){
  tinytest::test_package("attackerkb")
}



R/topics.R
#' Helpers to query AttackerKB topics
#'
#' The main `kb_topics` function maps 1:1 to the API. Leave values
#' `NULL` that you do not want included in the search parameters.
#'
#' Set the option `progress_enabled` to `FALSE` to disable progress spinner.
#'
#' @param topic_id UUID of a specific topic to return
#' @param editor_id UUID of a contributor
#' @param name Text to query the name attribute. A substring match is performed
#' @param created Return all topics that were created on the given date.
#' @param revised Return all topics that were revised on the given date.
#' @param disclosed Return all topics that were disclosed on the given date.
#' @param document Text to query the document attribute. A substring match is performed
#' @param metadata Text to query the metadata attribute. A substring match is performed
#' @param featured (lgl) `TRUE`/`FALSE`. Return all topics that are featured.
#' @param q Return all topics that have content that matches the query string.
#' @param api_key See [attackerkb_api_key()]
#' @references <https://api.attackerkb.com/api-docs/docs>
#' @export
kb_topics <- function(topic_id = NULL,
                      editor_id = NULL,
                      name = NULL,
                      created = NULL,
                      revised = NULL,
                      disclosed = NULL,
                      document = NULL,
                      metadata = NULL,
                      featured = NULL,
                      q = NULL,
                      api_key = attackerkb_api_key()) {

  topic_id <- topic_id[1]
  editor_id <-editor_id[1]
  name <- name[1]
  created <- created[1]
  revised <-revised[1]
  disclosed <- disclosed[1]
  document <- document[1]
  metadata <- metadata[1]
  featured <- featured[1]
  q <- q[1]

  if (length(featured)) featured <- tolower(as.character(as.logical(featured)))
  if (length(created)) created <- as.character(as.Date(created[1]))
  if (length(revised)) revised <- as.character(as.Date(revised[1]))
  if (length(disclosed)) disclosed <- as.character(as.Date(disclosed[1]))

  httr::GET(
    url = "https://api.attackerkb.com/topics",
    .ATTACKERKB_UA,
    query = list(
      id = topic_id,
      editorId = editor_id,
      name = name,
      created = created,
      revisionDate = revised,
      disclosureDate = disclosed,
      document = document,
      metadata = metadata,
      featured = featured,
      q = q,
      size = 500L
    ),
    httr::add_headers(`Authorization` = sprintf("basic %s", api_key))
  ) -> res

  httr::stop_for_status(res)

  out <- httr::content(res, as = "text", encoding = "UTF-8")
  out <- jsonlite::fromJSON(out)

  out <- handle_response(out, api_key = api_key, came_from = "kb_topics")

  out

}

#' @rdname kb_topics
#' @export
kb_topic <- function(topic_id = "131226a6-a1e9-48a1-a5d0-ac94baf8dfd2", api_key = attackerkb_api_key()) {

  httr::GET(
    url = sprintf("https://api.attackerkb.com/topics/%s", topic_id[1]),
    .ATTACKERKB_UA,
    httr::add_headers(`Authorization` = sprintf("basic %s", api_key))
  ) -> res

  httr::stop_for_status(res)

  out <- httr::content(res, as = "text", encoding = "UTF-8")
  out <- jsonlite::fromJSON(out)

  out <- handle_response(out, api_key = api_key, came_from = "kb_topic")

  out

}
#
# #' @rdname kb_topics
# #' @export
# kb_topics_by_contributor <- function(editor_id = "7191a637-aa4e-4885-98a0-f4f2da285b99", api_key = attackerkb_api_key()) {
#
#   httr::GET(
#     url = "https://api.attackerkb.com/topics",
#     .ATTACKERKB_UA,
#     query = list(
#       editorId = editor_id[1],
#       size = 500L
#     ),
#     httr::add_headers(`Authorization` = sprintf("basic %s", api_key))
#   ) -> res
#
#   httr::stop_for_status(res)
#
#   out <- httr::content(res, as = "text", encoding = "UTF-8")
#   out <- jsonlite::fromJSON(out)
#
#   out <- handle_response(out)
#
#   out
#
# }
#
# #' @rdname kb_topics
# #' @export
# kb_topics_by_name <- function(q = "bluekeep", api_key = attackerkb_api_key()) {
#
#   httr::GET(
#     url = "https://api.attackerkb.com/topics",
#     .ATTACKERKB_UA,
#     query = list(
#       name = q[1],
#       size = 500L
#     ),
#     httr::add_headers(`Authorization` = sprintf("basic %s", api_key))
#   ) -> res
#
#   httr::stop_for_status(res)
#
#   out <- httr::content(res, as = "text", encoding = "UTF-8")
#   out <- jsonlite::fromJSON(out)
#
#   out <- handle_response(out)
#
#   out
#
# }
#
# #' @rdname kb_topics
# #' @export
# kb_topics_by_document <- function(q = "bluekeep", api_key = attackerkb_api_key()) {
#
#   httr::GET(
#     url = "https://api.attackerkb.com/topics",
#     .ATTACKERKB_UA,
#     query = list(
#       document = q[1],
#       size = 500L
#     ),
#     httr::add_headers(`Authorization` = sprintf("basic %s", api_key))
#   ) -> res
#
#   httr::stop_for_status(res)
#
#   out <- httr::content(res, as = "text", encoding = "UTF-8")
#   out <- jsonlite::fromJSON(out)
#
#   out <- handle_response(out)
#
#   out
#
# }
#
#
# #' @rdname kb_topics
# #' @export
# kb_topics_by_content <- function(q = "bluekeep", api_key = attackerkb_api_key()) {
#
#   httr::GET(
#     url = "https://api.attackerkb.com/topics",
#     .ATTACKERKB_UA,
#     query = list(
#       content = q[1],
#       size = 500L
#     ),
#     httr::add_headers(`Authorization` = sprintf("basic %s", api_key))
#   ) -> res
#
#   httr::stop_for_status(res)
#
#   out <- httr::content(res, as = "text", encoding = "UTF-8")
#   out <- jsonlite::fromJSON(out)
#
#   out <- handle_response(out)
#
#   out
#
# }
#
# #' @rdname kb_topics
# #' @export
# kb_topics_by_metadata <- function(lookup = "bluekeep", api_key = attackerkb_api_key()) {
#
#   httr::GET(
#     url = "https://api.attackerkb.com/topics",
#     .ATTACKERKB_UA,
#     query = list(
#       metadata = lookup[1],
#       size = 500L
#     ),
#     httr::add_headers(`Authorization` = sprintf("basic %s", api_key))
#   ) -> res
#
#   httr::stop_for_status(res)
#
#   out <- httr::content(res, as = "text", encoding = "UTF-8")
#   out <- jsonlite::fromJSON(out)
#
#   out <- handle_response(out)
#
#   out
#
# }
#
#
# #' @rdname kb_topics
# #' @export
# kb_topics_by_date_created_date <- function(created = "2019-07-04", api_key = attackerkb_api_key()) {
#
#   created <- as.Date(created[1])
#
#   httr::GET(
#     url = "https://api.attackerkb.com/topics",
#     .ATTACKERKB_UA,
#     query = list(
#       created = as.character(created[1]),
#       size = 500L
#     ),
#     httr::add_headers(`Authorization` = sprintf("basic %s", api_key))
#   ) -> res
#
#   httr::stop_for_status(res)
#
#   out <- httr::content(res, as = "text", encoding = "UTF-8")
#   out <- jsonlite::fromJSON(out)
#
#   out <- handle_response(out)
#
#   out
#
# }
#
# #' @rdname kb_topics
# #' @export
# kb_topics_by_date_revised_date <- function(revised = "2019-07-04", api_key = attackerkb_api_key()) {
#
#   revised <- as.Date(revised[1])
#
#   httr::GET(
#     url = "https://api.attackerkb.com/topics",
#     .ATTACKERKB_UA,
#     query = list(
#       revisionDate = as.character(revised[1]),
#       size = 500L
#     ),
#     httr::add_headers(`Authorization` = sprintf("basic %s", api_key))
#   ) -> res
#
#   httr::stop_for_status(res)
#
#   out <- httr::content(res, as = "text", encoding = "UTF-8")
#   out <- jsonlite::fromJSON(out)
#
#   out <- handle_response(out)
#
#   out
#
# }
#
# #' @rdname kb_topics
# #' @export
# kb_topics_by_date_disclosed_date <- function(disclosed = "2019-07-04", api_key = attackerkb_api_key()) {
#
#   disclosed <- as.Date(disclosed[1])
#
#   httr::GET(
#     url = "https://api.attackerkb.com/topics",
#     .ATTACKERKB_UA,
#     query = list(
#       disclosureDate = as.character(disclosed[1]),
#       size = 500L
#     ),
#     httr::add_headers(`Authorization` = sprintf("basic %s", api_key))
#   ) -> res
#
#   httr::stop_for_status(res)
#
#   out <- httr::content(res, as = "text", encoding = "UTF-8")
#   out <- jsonlite::fromJSON(out)
#
#   out <- handle_response(out)
#
#   out
#
# }
#


R/api-key.R
#' Get or set ATTACKERKB_API_KEY value
#'
#' The API wrapper functions in this package all rely on a AttackerKB API
#' key residing in the environment variable `ATTACKERKB_API_KEY`.
#' The easiest way to accomplish this is to set it
#' in the `.Renviron` file in your home directory.
#'
#' You can obtain an AttackerKB API key by going visiting
#' **AttackerKB Profile Page > Settings > API Key.**
#'
#' @md
#' @param force Force setting a new AttackerKB key for the current environment?
#' @return atomic character vector containing the AttackerKB API key
#' @export
attackerkb_api_key <- function(force = FALSE) {

  env <- Sys.getenv('ATTACKERKB_API_KEY')
  if (!identical(env, "") && !force) return(env)

  if (!interactive()) {
    stop("Please set env var ATTACKERKB_API_KEY to your AttackerKB key",
         call. = FALSE)
  }

  message("Couldn't find env var ATTACKERKB_API_KEY See ?attackerkb_api_key for more details.")
  message("Please enter your API key:")
  pat <- readline(": ")

  if (identical(pat, "")) {
    stop("AttackerKB key entry failed", call. = FALSE)
  }

  message("Updating ATTACKERKB_API_KEY env var")
  Sys.setenv(ATTACKERKB_API_KEY = pat)

  pat

}


R/assessments.R
#' Helpers to query AttackerKB assessments
#'
#' The main `kb_contributors` function maps 1:1 to the API. Leave values
#' `NULL` that you do not want included in the search parameters.
#'
#' Set the option `progress_enabled` to `FALSE` to disable progress spinner.
#'
#' @param assessment_id UUID of a specific assessment to return
#' @param editor_id  UUID of a contributor.
#' @param topid_id UUID of the topic this assessment was based on.
#' @param created Return all assessments that were created on the given date.
#' @param revised Return all assessments that were revised on the given date.
#' @param document Text to query the document attribute. A substring match is performed
#' @param metadata Text to query the metadata attribute. A substring match is performed
#' @param score Return all assessments with this score.
#' @param q Return all assessments that have content that match the query string.
#' @param api_key See [attackerkb_api_key()]
#' @references <https://api.attackerkb.com/api-docs/docs>
#' @export
kb_assessments <- function(assessment_id = NULL,
                           editor_id = NULL,
                           topid_id = NULL,
                           created = NULL,
                           revised = NULL,
                           document = NULL,
                           metadata = NULL,
                           score = NULL,
                           q = NULL,
                           api_key = attackerkb_api_key()) {

  assessment_id <- assessment_id[1]
  editor_id <- editor_id[1]
  topid_id <- topid_id[1]
  created <- created[1]
  revised <-revised[1]
  document <- document[1]
  metadata <- metadata[1]
  score <-score[1]
  q <- q[1]

  if (length(created)) created <- as.character(as.Date(created[1]))
  if (length(revised)) created <- as.character(as.Date(created[1]))

  httr::GET(
    url = "https://api.attackerkb.com/assessments",
    .ATTACKERKB_UA,
    query = list(
      id = assessment_id,
      editorId = editor_id,
      topicId = topid_id,
      created = created,
      revisionDate = revised,
      document = document,
      metadata = metadata,
      score = score,
      q = q,
      size = 500L
    ),
    httr::add_headers(`Authorization` = sprintf("basic %s", api_key))
  ) -> res

  httr::stop_for_status(res)

  out <- httr::content(res, as = "text", encoding = "UTF-8")
  out <- jsonlite::fromJSON(out)

  out <- handle_response(out, api_key = api_key, came_from = "kb_assessments")

  out

}

#' @rdname kb_assessments
#' @export
kb_assessment <- function(assessment_id, api_key = attackerkb_api_key()) {

  assessment_id <- assessment_id[1]

  httr::GET(
    url = sprintf("https://api.attackerkb.com/assessments/%s", assessment_id),
    .ATTACKERKB_UA,
    httr::add_headers(`Authorization` = sprintf("basic %s", api_key))
  ) -> res

  httr::stop_for_status(res)

  out <- httr::content(res, as = "text", encoding = "UTF-8")
  out <- jsonlite::fromJSON(out)

  out <- out$data

  metadata <- out$metadata
  out$metadata <- NULL

  out <- as.data.frame(out, stringsAsFactors = FALSE)

  out$tags <- list(metadata[["tags"]]) %l0% NA_character_
  out$attacker_value <- metadata[["attacker-value"]] %l0% NA_integer_
  out$exploitability <- metadata[["exploitability"]] %l0% NA_integer_
  out$stability <- metadata[["stability"]] %l0% NA_integer_
  out$reliability <- metadata[["reliability"]] %l0% NA_integer_
  out$urgent_to_patch <- metadata[["urgent-to-patch"]] %l0% NA_integer_
  out$used_successfully <- metadata[["used-successfully"]] %l0% NA_integer_
  out$mitigation_strength <- metadata[["mitigation-strength"]] %l0% NA_integer_
  out$confidence_in_ratings <- metadata[["confidence-in-ratings"]] %l0% NA_integer_
  out$effort_to_develop_exploit <- metadata[["effort-to-develop-exploit"]] %l0% NA_integer_
  out$versions <- list(metadata[["versions"]]) %||% list()
  out$mitigation <- metadata[["mitigation"]] %l0% NA_character_
  out$forever_day_versions <- list(metadata[["forever_day_versions"]]) %l0% list()
  out$offensive_application <- metadata[["offensive-application"]] %l0% NA_character_
  out$atacker_utility <- metadata[["atacker-utility"]] %l0% NA_integer_
  out$urgent_topatch <- metadata[["urgent-to patch"]] %l0% NA_integer_

  class(out) <- c("tbl_df", "tbl", "data.frame")

  out


}

R/contributor.R
#' Helpers to query AttackerKB contributors
#'
#' The main `kb_contributors` function maps 1:1 to the API. Leave values
#' `NULL` that you do not want included in the search parameters.
#'
#' Set the option `progress_enabled` to `FALSE` to disable progress spinner.
#'
#' @param contributor_id UUID of a specific contributror to return
#' @param username Return contributors with the matching username.
#' @param avatar Return all contributors where avatar matches the given value
#' @param created Return all contributors that were created on the given date.
#' @param score Return all contributors with this score.
#' @param q Return all contributors that have usernames that match the query string.
#' @param api_key See [attackerkb_api_key()]
#' @references <https://api.attackerkb.com/api-docs/docs>
#' @export
kb_contributors <- function(contributor_id = NULL,
                            username = NULL,
                            avatar = NULL,
                            created = NULL,
                            score = NULL,
                            q = NULL,
                            api_key = attackerkb_api_key()) {

  contributor_id <- contributor_id[1]
  username <- username[1]
  avatar <- avatar[1]
  created <- created[1]
  score <- score[1]
  q <- q[1]

  if (length(created)) created <- as.character(as.Date(created[1]))

  httr::GET(
    url = "https://api.attackerkb.com/contributors",
    .ATTACKERKB_UA,
    query = list(
      id = contributor_id,
      username = username,
      avatar = avatar,
      created = created,
      score = score,
      q = q,
      size = 500L
    ),
    httr::add_headers(`Authorization` = sprintf("basic %s", api_key))
  ) -> res

  httr::stop_for_status(res)

  out <- httr::content(res, as = "text", encoding = "UTF-8")
  out <- jsonlite::fromJSON(out)

  out <- handle_response(out, api_key = api_key, came_from = "kb_contributors")

  out

}

#' @rdname kb_contributors
#' @export
kd_contributor <- function(contributor_id, api_key = attackerkb_api_key()) {

  contributor_id <- contributor_id[1]

  httr::GET(
    url = sprintf("https://api.attackerkb.com/contributors/%s", contributor_id),
    .ATTACKERKB_UA,
    httr::add_headers(`Authorization` = sprintf("basic %s", api_key))
  ) -> res

  httr::stop_for_status(res)

  out <- httr::content(res, as = "text", encoding = "UTF-8")
  out <- jsonlite::fromJSON(out)

  out <- as.data.frame(out$data, stringsAsFactors=FALSE)

  class(out) <- c("tbl_df", "tbl", "data.frame")
  out


}

R/utils.R
date_convert <- function(.x) {
  if (length(.x)) {
    as.POSIXct(.x, format="%Y-%m-%dT%H:%M:%OS", tz="GMT")
  } else {
    as.POSIXct(NA)
  }
}

.kb_reshape <- function(.x, path) {

  if (length(.x) == 0) return(data.frame(stringsAsFactors=FALSE))

  if (path == "/topics") {

    easy_cols <- .x[, c("id", "editorId", "name", "created", "revisionDate", "disclosureDate", "document")]

    easy_cols[["created"]] <- date_convert(easy_cols[["created"]])
    easy_cols[["revisionDate"]] <- date_convert(easy_cols[["revisionDate"]])

    easy_cols$references <- .x[["metadata"]][["references"]]

    cvss_v3 <- .x[["metadata"]][["baseMetricV3"]][["cvssV3"]]

    easy_cols$cvss_v3_scope <- cvss_v3[["scope"]]
    easy_cols$cvss_v3_version <- cvss_v3[["version"]]
    easy_cols$cvss_v3_baseScore <- cvss_v3[["baseScore"]]
    easy_cols$cvss_v3_attackVector <- cvss_v3[["attackVector"]]
    easy_cols$cvss_v3_baseSeverity <- cvss_v3[["baseSeverity"]]
    easy_cols$cvss_v3_vectorString <- cvss_v3[["vectorString"]]
    easy_cols$cvss_v3_integrityImpact <- cvss_v3[["integrityImpact"]]
    easy_cols$cvss_v3_userInteraction <- cvss_v3[["userInteraction"]]
    easy_cols$cvss_v3_attackComplexity <- cvss_v3[["attackComplexity"]]
    easy_cols$cvss_v3_availabilityImpact <- cvss_v3[["availabilityImpact"]]
    easy_cols$cvss_v3_privilegesRequired <- cvss_v3[["privilegesRequired"]]
    easy_cols$cvss_v3_confidentialityImpact <- cvss_v3[["confidentialityImpact"]]
    easy_cols$cvss_v3_impact <- .x[["metadata"]][["baseMetricV3"]][["impactScore"]]
    easy_cols$cvss_v3_exploitability <- .x[["metadata"]][["baseMetricV3"]][["exploitabilityScore"]]
    easy_cols$vulnerable_versions <- .x[["metadata"]][["vulnerable-versions"]]
    easy_cols$kb_attacker_value <- .x[["score"]][["attackerValue"]]
    easy_cols$kb_exploitability <- .x[["score"]][["exploitability"]]
    easy_cols$tag_commonEnterprise <- .x$tags[["commonEnterprise"]]
    easy_cols$tag_defaultConfiguration <- .x$tags[["defaultConfiguration"]]
    easy_cols$tag_difficultToPatch <- .x$tags[["difficultToPatch"]]
    easy_cols$tag_highPrivilegeAccess <- .x$tags[["highPrivilegeAccess"]]
    easy_cols$tag_easyToDevelop <- .x$tags[["easyToDevelop"]]
    easy_cols$tag_requiresInteraction <- .x$tags[["requiresInteraction"]]
    easy_cols$tag_obscureConfiguration <- .x$tags[["obscureConfiguration"]]
    easy_cols$tag_difficultToExploit <- .x$tags[["difficultToExploit"]]
    easy_cols$tag_noUsefulData <- .x$tags[["noUsefulData"]]
    easy_cols$tag_difficultToDevelop <- .x$tags[["difficultToDevelop"]]
    easy_cols$tag_preAuth <- .x$tags[["preAuth"]]
    easy_cols$tag_postAuth <- .x$tags[["postAuth"]]

    easy_cols

    rownames(easy_cols) <- NULL

    easy_cols

  } else if (path == "/topic") {

    ## List of 10
    ## $ created       : chr "2019-05-14T18:28:19.31074Z"
    ## $ disclosureDate: chr "2019-05-16T19:29:00Z"
    ## $ document      : chr "A bug in Windows Remote Desktop protocol allows unauthenticated users to run arbitrary code via a specially cra"| __truncated__
    ## $ editorId      : chr "7191a637-aa4e-4885-98a0-f4f2da285b99"
    ## $ id            : chr "131226a6-a1e9-48a1-a5d0-ac94baf8dfd2"
    ## $ metadata      :List of 16
    ## $ name          : chr "Windows Remote Desktop (RDP) Use-after-free vulnerablility, \"Bluekeep\""
    ## $ revisionDate  : chr "2020-03-03T16:18:02.56368Z"
    ## $ score         :List of 2
    ## $ tags          :List of 12

    .x[["created"]] <- date_convert(.x[["created"]])
    .x[["disclosureDate"]] <- date_convert(.x[["disclosureDate"]])
    .x[["metadata"]] <- I(list(.x[["metadata"]]))
    .x[["score"]] <- I(list(.x[["score"]]))
    .x[["tags"]] <- I(list(.x[["tags"]]))

    if ("rapid7Analysis" %in% names(.x)) {
      if (length(.x[["rapid7Analysis"]])) {
        .x[["rapid7Analysis"]] <- paste0(.x[["rapid7Analysis"]], collapse = " ")
      } else {
        .x[["rapid7Analysis"]] <- NA_character_
      }
    }

    .x <- as.data.frame(.x, stringsAsFactors = FALSE)

    .x

  } else if (path == "/contributors") {

    .x[["created"]] <- date_convert(.x[["created"]])
    .x

  } else if (path == "/assessments") {

    easy_cols <- .x[,c("id", "editorId", "topicId", "created", "revisionDate", "document", "score")]

    easy_cols[["created"]] <- date_convert(easy_cols[["created"]])
    easy_cols[["revisionDate"]] <- date_convert(easy_cols[["revisionDate"]])

    metadata <- easy_cols[["metadata"]]

    easy_cols$tags <- metadata[["tags"]]
    easy_cols$attacker_value <- metadata[["attacker-value"]]
    easy_cols$exploitability <- metadata[["exploitability"]]
    easy_cols$stability <- metadata[["stability"]]
    easy_cols$reliability <- metadata[["reliability"]]
    easy_cols$urgent_to_patch <- metadata[["urgent-to-patch"]]
    easy_cols$used_successfully <- metadata[["used-successfully"]]
    easy_cols$mitigation_strength <- metadata[["mitigation-strength"]]
    easy_cols$confidence_in_ratings <- metadata[["confidence-in-ratings"]]
    easy_cols$effort_to_develop_exploit <- metadata[["effort-to-develop-exploit"]]
    easy_cols$versions <- metadata[["versions"]]
    easy_cols$mitigation <- metadata[["mitigation"]]
    easy_cols$forever_day_versions <- metadata[["forever_day_versions"]]
    easy_cols$offensive_application <- metadata[["offensive-application"]]
    easy_cols$atacker_utility <- metadata[["atacker-utility"]]
    easy_cols$urgent_topatch <- metadata[["urgent-to patch"]]

    easy_cols

  }

}

handle_response <- function(.x, api_key = attackerkb_api_key(), came_from = NULL) {

  .pb <- progress::progress_bar$new(format = "(:spin)", total = NA)
  .pb$tick()

  path <- .x[["links"]][["self"]][["href"]]
  if (length(path) == 0) path <- sub("^kb_", "/", came_from)

  ret <- .x[["data"]]

  ret <- .kb_reshape(ret, path)

  next_href <- .x[["links"]][["next"]][["href"]]

  while(length(next_href)) {

    .pb$tick()

    httr::GET(
      url = sprintf("https://api.attackerkb.com%s", next_href),
      .ATTACKERKB_UA,
      httr::add_headers(`Authorization` = sprintf("basic %s", api_key))
    ) -> res

    httr::stop_for_status(res)

    out <- httr::content(res, as = "text", encoding = "UTF-8")
    out <- jsonlite::fromJSON(out)

    next_href <- out[["links"]][["next"]][["href"]]

    ret <- data.table::rbindlist(list(ret, .kb_reshape(out[["data"]], path)), fill = TRUE)

  }

  class(ret) <- c("tbl_df", "tbl", "data.frame")
  ret

}
